### PR TITLE
Add cpu topology settings and possibility to overcommit guest overhead

### LINKED ISSUE #36: Add additional VM configuration options
Currently it's already possible to specify essential VM parameters such as CPUs, memory, disk size, and storage class for the VM disk. We should be able to specify also additional VM parameters, usually for an enhanced VM performance. These could include, among others:

* [hugepages support](https://kubevirt.io/user-guide/#/installation/hugepages)
* [dedicated CPU](https://kubevirt.io/user-guide/#/creation/dedicated-cpu)
* [disk device cache](https://kubevirt.io/user-guide/#/creation/disks-and-volumes?id=disk-device-cache)

We should discuss what additional parameters should be added, and how should they be configured. The options are:

* In the `spec.providerConfig` section of the cloud profile, if to configure machine types shared between multiple shoots
* In the `provider.workers[n].providerConfig` section of the shoot resource, to configure individual shoots.

/area performance
/kind enhancement
/priority normal
/platform kubevirt

### PR DESCRIPTION
**Pull request by mfranczy**
Signed-off-by: Marcin Franczyk <marcin0franczyk@gmail.com>

**How to categorize this PR?**

/area performance
/kind enhancement
/priority normal
/platform kubevirt

**What this PR does / why we need it**:
It allows to set CPU topology and supports memory overcommit

**Which issue(s) this PR fixes**:
Fixes #36 

**Special notes for your reviewer**:

**Release note**:

```improvement operator
improvement_user
```

### CODE REVIEW

**Comment by mfranczy** [comment]
Please do not merge yet, that change breaks API. It will be possible to merge this when I am done with MCM provider kubevirt.

File: docs/usage-as-end-user.md
Comment (stoyanr) [inline docs/usage-as-end-user.md:140]:
```suggestion
  # features specifies the CPU features list inside the VMI
```

File: docs/usage-as-operator.md
Comment (stoyanr) [inline docs/usage-as-operator.md:27]: Could you please change this API as follows:
```yaml
machineTypes:
- name: standard-1
  limits:
    cpu: "2"
    memory: 8Gi
```
I think it's likely that we would need to specify additional kubevirt-specific info for machine types in the future, so let's have one section `machineTypes` instead of `machineTypesLimits` and then `machineTypesWhatever`.

File: pkg/apis/kubevirt/types_cloudprofile.go
Comment (stoyanr) [inline pkg/apis/kubevirt/types_cloudprofile.go:32]: Please change this to `MachineTypes` and add the limits inside, see above.

File: pkg/apis/kubevirt/v1alpha1/types_cloudprofile.go
Comment (stoyanr) [inline pkg/apis/kubevirt/v1alpha1/types_cloudprofile.go:33]: Please change this to `MachineTypes` and add the limits inside, see above.

File: charts/internal/machine-class/templates/machine-class.yaml
Comment (stoyanr) [inline charts/internal/machine-class/templates/machine-class.yaml:59]: You already have a field named `memory` above. I believe `cpus` and that old `memory` field are no longer needed and should be replaced by `resources`?

File: charts/internal/machine-class/values.yaml
Comment (stoyanr) [inline charts/internal/machine-class/values.yaml:23]: You already have a field named `memory` above. I believe `cpus` and that old `memory` field are no longer needed and should be replaced by `resources`?

File: pkg/controller/worker/machines.go
Comment (stoyanr) [inline pkg/controller/worker/machines.go:166]: Perhaps this could be renamed to `resources`, as in Kubernetes?

**Review by stoyanr — CHANGES_REQUESTED** [review]
Nice PR, see comments for some issues / requests.

File: docs/usage-as-operator.md
Comment (mfranczy) [reply docs/usage-as-operator.md:27]: good idea.

File: charts/internal/machine-class/templates/machine-class.yaml
Comment (mfranczy) [reply charts/internal/machine-class/templates/machine-class.yaml:59]: good catch, and yes I can change it to resources.

**Comment by mfranczy** [comment]
as we spoke I will rebase against master and add validation

File: pkg/apis/kubevirt/v1alpha1/types_worker.go
Comment (stoyanr) [inline pkg/apis/kubevirt/v1alpha1/types_worker.go:31]:
```suggestion
	// +optional
	CPU *kubevirtv1.CPU `json:"cpu,omitempty"`
```

File: pkg/apis/kubevirt/v1alpha1/types_cloudprofile.go
Comment (stoyanr) [inline pkg/apis/kubevirt/v1alpha1/types_cloudprofile.go:54]: The name of this type should be `MachineType`, as it refers to a single machine type. `MachineImages` is similarly wrong and could be renamed to `MachineImage` as well.

File: pkg/apis/kubevirt/types_cloudprofile.go
Comment (stoyanr) [inline pkg/apis/kubevirt/types_cloudprofile.go:52]: The name of this type should be `MachineType`, as it refers to a single machine type. `MachineImages` is similarly wrong and could be renamed to `MachineImage` as well.

File: pkg/controller/worker/machine_config.go
Comment (stoyanr) [inline pkg/controller/worker/machine_config.go:86]: Checking a slice for nil is not needed, a nil slice is an empty slice, and is iterable.
```suggestion
	if w.cloudProfileConfig != nil {
```

File: pkg/controller/worker/machines.go
Comment (stoyanr) [inline pkg/controller/worker/machines.go:152]: Checking a slice for nil is not needed, a nil slice is an empty slice. Also, this case should be handled by `getMachineTypesExtension` so no need to do double work here.

**Review by stoyanr — CHANGES_REQUESTED** [review]
Thanks for addressing the previous comments, adding a few more minor things.

File: pkg/apis/kubevirt/v1alpha1/types_worker.go
Comment (mfranczy) [reply pkg/apis/kubevirt/v1alpha1/types_worker.go:32]: done

File: pkg/apis/kubevirt/v1alpha1/types_cloudprofile.go
Comment (mfranczy) [reply pkg/apis/kubevirt/v1alpha1/types_cloudprofile.go]: `MachineType` is incorrect I agree, but when it comes to `MachineImages` it's correct.
`MachineImages` structure can contain more than one image (list of versions with source URL).

Although it requires some refactoring, the part about getting machine images has been copy-pasted from the vsphere provider, I think it can be simplified.  Currently we have two `MachineImages(s)` structures, in worker and cloud profile files, we could just use one. I am going to create an issue for that were I explain the problem and how to solve that.

File: pkg/apis/kubevirt/types_cloudprofile.go
Comment (mfranczy) [reply pkg/apis/kubevirt/types_cloudprofile.go]: https://github.com/gardener/gardener-extension-provider-kubevirt/pull/72#discussion_r500246304

File: pkg/controller/worker/machine_config.go
Comment (mfranczy) [reply pkg/controller/worker/machine_config.go]: done

File: pkg/controller/worker/machines.go
Comment (mfranczy) [reply pkg/controller/worker/machines.go]: done

File: pkg/apis/kubevirt/v1alpha1/types_cloudprofile.go
Comment (stoyanr) [reply pkg/apis/kubevirt/v1alpha1/types_cloudprofile.go]: OK, agreed. Let's discuss the `MachineImages` simplification in person before opening issues or making changes.

**Review by stoyanr — COMMENTED** [review]
Thanks for addressing the feedback so far, no more comments, waiting to see validation before approving.

**Comment by mfranczy** [comment]
@stoyanr I will resolve conflicts and not to block you, let's merge it and I will open another PR with validation

**Comment by mfranczy** [comment]
@stoyanr rebase is done.

**Review by stoyanr — APPROVED** [review]

**Pull request merged**
